src/statement.ts

```diff
@@ -0,0 +1,240 @@
+            type,
+            tables,
+          });
+          const results = await queryAll<T>(config, sql, opts);
+          if (colName != null) {
+            return wrapResult(
+              extractColumn(results, colName),
+              performance.now() - start
+            );
+          }
+          return wrapResult(results, performance.now() - start);
+        }
+        default: {
+          let receipt = await exec(this.config, { type, sql, tables });
+          if (this.config.autoWait ?? false) {
+            const waited = await receipt.wait();
+            receipt = { ...receipt, ...waited };
+          }
+
+          return wrapResult(receipt, performance.now() - start);
+        }
+      }
+    } catch (cause) {
+      if (
+        cause instanceof Error &&
+        cause.message.includes("column not found")
+      ) {
+        throw new Error("COLUMN_NOTFOUND", { cause });
+      }
+      throw new Error("ALL_ERROR", { cause });
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async first<T = S, K extends keyof T = keyof T>(): Promise<T>;
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async first<T = S, K extends keyof T = keyof T>(
+    colName: undefined,
+    opts?: Signal
+  ): Promise<T>;
+  async first<T = S, K extends keyof T = keyof T>(
+    colName: K,
+    opts?: Signal
+  ): Promise<T[K] | null>;
+  async first<T = S, K extends keyof T = keyof T>(
+    colName?: K,
+    opts: Signal = {}
+  ): Promise<T | T[K] | null> {
+    try {
+      const { sql, type, tables } = await this.#parseAndExtract();
+      switch (type) {
+        case "read": {
+          const config = await extractReadonly(this.config, {
+            type,
+            tables,
+          });
+          const results = await queryFirst<T>(config, sql, opts);
+          if (results == null || colName == null) {
+            return results;
+          }
+          return extractColumn(results, colName);
+        }
+        default: {
+          const receipt = await exec(this.config, { type, sql, tables });
+          /* c8 ignore next */
+          if (this.config.autoWait ?? false) {
+            await receipt.wait();
+          }
+          return null;
+        }
+      }
+    } catch (cause) {
+      if (
+        cause instanceof Error &&
+        cause.message.includes("column not found")
+      ) {
+        throw new Error("COLUMN_NOTFOUND", { cause });
+      }
+      throw new Error("FIRST_ERROR", { cause });
+    }
+  }
+
+  async run(opts: Signal = {}): Promise<Result<never>> {
+    try {
+      const start = performance.now();
+      const { sql, type, tables } = await this.#parseAndExtract();
+      switch (type) {
+        case "read": {
+          const config = await extractReadonly(this.config, {
+            type,
+            tables,
+          });
+          await queryAll(config, sql, opts);
+          return wrapResult([], performance.now() - start);
```

Comment:
I don't think, if they've read the docs correctly, anyone would use run with a read query. But since it is possible for them to do it, I went with this. This is the behavior that D1 follows. It'll still run the query, but it doesn't return the result. I guess it might be useful for timing the response? Not really sure.
- I don't think logging a warning is needed, since this behavior is already documented.
- I don't like the idea of adding an extra parameter here, just for this.
- But there _will_ be results in the case that the read query was valid. When we call wrap results, it'll take that and include it in the output.

Anyway, I'm ok with us just _not_ throwing away the returned response in this case (and including it as the "results" object), just because it seems slightly more intuitive if you did give a ready query to the run statement. The only downside is that this does _not_ match the D1 behavior. I guess that's ok? As it can simply be ignored by downstream consumers. I get that it seems silly to even expect a response at all in this case.

Reply:
Thinking about this a bit more, I still don't think we want an extra flag here on the API, because we only have a read API. So there is really no point in calling it if you don't want results. I had originally throw an error when you tried to run a read query, but D1 does actually execute the query it seems, and returns an empty result. So I went with that. But we can just include the results here if that feels more intuitive. Long story short though: don't use run for reads.

Reply:
I guess leaving it as is makes the most sense, good docs will make this a non-issue.

Reply:
Well I went and made it so that it'll return the results if it is a read. I think that's fine. No sense in wasting an API call right? If it turns out to be a compatibility issue, we can revert. But we'll make sure the docs say the right thing (which the API.md doc does).